### ISSUE TITLE
[apex] ApexSuggestUsingNamedCred false positive with Named Credential merge fields

### ISSUE THREAD

**Issue by erik4github** [pmd/pmd#3158]
**Affects PMD Version:**
6.32.0+
7.0.0

**Rule:** ApexSuggestUsingNamedCred

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_apex_security.html#apexsuggestusingnamedcred>

**Description:** Warning is "Consider using named credentials for authenticated callouts (rule: Security-ApexSuggestUsingNamedCred)apex pmdApexSuggestUsingNamedCred", but I am using Named Credentials with merge fields, the same way the official Salesforce docs list them: https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_callouts_named_credentials_merge_fields.htm

**Code Sample demonstrating the issue:**

```apex
  private static HttpRequest buildHTTPRequest(String endpoint, String httpMethod, String body) {
    HttpRequest req = new HttpRequest();
    req.setEndpoint(endpoint);
    req.setMethod(httpMethod);
    req.setHeader('Authorization', '{!$Credential.AuthorizationMethod} {!$Credential.Password}');
    req.setHeader('Content-Type', 'application/json');
    req.setBody(body);

    return req;
  }
```

**Expected outcome:**

*   Does PMD report a violation, where there shouldn't be one? -> false-positive

`req.setHeader('Authorization', '{!$Credential.AuthorizationMethod} {!$Credential.Password}');` shouldn't suggest using Named Credentials because it is using Named Credentials, through merge fields.

**Running PMD through:** *Visual Studio Code*

**Comment by adangel** [comment]
Yes, I think the rule needs to be updated to supported Named Credential merge fields.
This feature seems to be not new, in the doc it's available since Winter '16 (API version 35).

With using Named Credential merge fields, there are no hardcoded credentials , so this is definitely a false positive.

**Comment by Damecek** [comment]
I have the same issue. Following code will throw warning on line `sms.applicationx5ftoken = '{!$Credential.Password}'`. This definitely is false positive. This is the [class](pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexSuggestUsingNamedCredRule.java) which is behind this. There should be some exceptions, for example this `{!$Credential.Password}` is standardized usage coming directly from Salesforce documentation.

```apex
private SMS() {
    sms = new ExternalService.BulkGate_SMSRequest();
    sms.applicationx5fid = '{!$Credential.Username}';
    sms.applicationx5ftoken = '{!$Credential.Password}';
}
```

![image](https://github.com/pmd/pmd/assets/46635736/b6d4ef38-c115-4302-9914-16a50912e953)
